File: components/jpa-handling-session/src/main/java/com/sv/youapp/component/authorization/services/impl/JpaRegisteredClientRepository.java
package com.sv.youapp.component.authorization.services.impl;

import com.sv.youapp.component.authorization.entities.jpa.RegisteredClientEntity;
import com.sv.youapp.component.authorization.repositories.jpa.ClientRepository;
import org.springframework.security.oauth2.core.AuthorizationGrantType;
import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository;
import org.springframework.security.oauth2.server.authorization.settings.ClientSettings;
import org.springframework.security.oauth2.server.authorization.settings.OAuth2TokenFormat;
import org.springframework.security.oauth2.server.authorization.settings.TokenSettings;

import java.util.Locale;
import java.util.Optional;

public record JpaRegisteredClientRepository(ClientRepository clientRepository) implements RegisteredClientRepository {

	@Override
	public void save(RegisteredClient registeredClient) {
		throw new UnsupportedOperationException("Not supported yet.");
	}

	@Override
	public RegisteredClient findById(String id) {
		if (id == null || id.isBlank()) return null;
		Optional<RegisteredClientEntity> opt = clientRepository.findById(id);
		return opt.map(JpaRegisteredClientRepository::toRegisteredClient).orElse(null);
	}

	@Override
	public RegisteredClient findByClientId(String clientId) {
		if (clientId == null || clientId.isBlank()) return null;
		Optional<RegisteredClientEntity> opt = clientRepository.findByClientId(clientId);
		return opt.map(JpaRegisteredClientRepository::toRegisteredClient).orElse(null);
	}

	private static RegisteredClient toRegisteredClient(RegisteredClientEntity self) {
		RegisteredClient.Builder builder = RegisteredClient.withId(self.getId())
			.clientId(self.getClientId())
			.clientSecret(self.getClientSecret())
			.clientName(self.getClientName())
			.clientIdIssuedAt(self.getClientIdIssuedAt())
			.clientSecretExpiresAt(self.getClientSecretExpiresAt());

		// Authentication methods
		self.getAuthenticationMethods().forEach(method ->
			builder.clientAuthenticationMethod(new ClientAuthenticationMethod(method.getName()))
		);

		// Grant types
		self.getGrantTypes().forEach(grant ->
			builder.authorizationGrantType(new AuthorizationGrantType(grant.getName()))
		);

		// Redirect URIs
		self.getRedirectUris().forEach(uri ->
			builder.redirectUri(uri.getRedirectUri())
		);

		// Post-logout Redirect URIs
		self.getPostLogoutRedirectUris().forEach(uri ->
			builder.postLogoutRedirectUri(uri.getRedirectUri())
		);

		// Scopes
		self.getScopes().forEach(scope ->
			builder.scope(scope.getName())
		);

		if (self.getClientSettings() != null) {
			var cs = self.getClientSettings();
			ClientSettings.Builder csBuilder = ClientSettings.builder();
			csBuilder.requireProofKey(cs.getRequireProofKey());
			csBuilder.requireAuthorizationConsent(cs.getRequireAuthorizationConsent());
			builder.clientSettings(csBuilder.build());
		}

		if (self.getTokenSettings() != null) {
			var ts = self.getTokenSettings();
			TokenSettings.Builder tsBuilder = TokenSettings.builder();

			tsBuilder.accessTokenTimeToLive(ts.getAccessTokenTimeToLive());
			tsBuilder.reuseRefreshTokens(false);
			tsBuilder.refreshTokenTimeToLive(ts.getRefreshTokenTimeToLive());

			String fmt = ts.getAccessTokenFormat();
			if (fmt != null) {
				String f = fmt.toLowerCase(Locale.ROOT);
				if (f.equals("reference")) {
					tsBuilder.accessTokenFormat(OAuth2TokenFormat.REFERENCE);
				} else {
					tsBuilder.accessTokenFormat(OAuth2TokenFormat.SELF_CONTAINED);
				}
			}
			builder.tokenSettings(tsBuilder.build());
		}
		return builder.build();
	}
}

File: components/jpa-handling-session/src/main/kotlin/com/sv/youapp/component/authorization/services/impl/JpaRegisteredClientRepository.kt
package com.sv.youapp.component.authorization.services.impl

import com.sv.youapp.component.authorization.entities.jpa.RegisteredClientEntity
import com.sv.youapp.component.authorization.repositories.jpa.ClientRepository
import org.springframework.security.oauth2.core.AuthorizationGrantType
import org.springframework.security.oauth2.core.ClientAuthenticationMethod
import org.springframework.security.oauth2.server.authorization.client.RegisteredClient
import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository
import org.springframework.security.oauth2.server.authorization.settings.ClientSettings
import org.springframework.security.oauth2.server.authorization.settings.OAuth2TokenFormat
import org.springframework.security.oauth2.server.authorization.settings.TokenSettings

class JpaRegisteredClientRepository
(private val clientRepository: ClientRepository) : RegisteredClientRepository {
    override fun save(registeredClient: RegisteredClient?) {
        // TODO("NOT IMPLEMENTED")
    }

    override fun findById(id: String?): RegisteredClient? {
        if (id.isNullOrBlank()) return null
        return clientRepository.findById(id)
            .map { it.toRegisteredClient() }
            .orElse(null)
    }

    override fun findByClientId(clientId: String?): RegisteredClient? {
        if (clientId.isNullOrBlank()) return null
        return clientRepository.findByClientId(clientId)
            .map { it.toRegisteredClient() }
            .orElse(null)
    }
}

private fun RegisteredClientEntity.toRegisteredClient(): RegisteredClient {
    val builder =
        RegisteredClient.withId(this.id)
            .clientId(this.clientId)
            .clientSecret(this.clientSecret)
            .clientName(this.clientName)
            .clientIdIssuedAt(this.clientIdIssuedAt)
            .clientSecretExpiresAt(this.clientSecretExpiresAt)

    // Authentication methods
    this.authenticationMethods.forEach { method ->
        builder.clientAuthenticationMethod(ClientAuthenticationMethod(method.name))
    }

    // Grant types
    this.grantTypes.forEach { grant ->
        builder.authorizationGrantType(AuthorizationGrantType(grant.name))
    }

    // Redirect URIs
    this.redirectUris.forEach { uri ->
        builder.redirectUri(uri.redirectUri)
    }
    // Post URIs
    this.postLogoutRedirectUris.forEach { uri ->
        builder.postLogoutRedirectUri(uri.redirectUri)
    }

    // Scopes
    this.scopes.forEach { scope ->
        builder.scope(scope.name)
    }

    this.clientSettings?.let { cs ->
        val csBuilder = ClientSettings.builder()
        cs.requireProofKey.let { csBuilder.requireProofKey(it) }
        cs.requireAuthorizationConsent.let { csBuilder.requireAuthorizationConsent(it) }
        builder.clientSettings(csBuilder.build())
    }

    this.tokenSettings?.let { ts ->
        val tsBuilder = TokenSettings.builder()

        ts.accessTokenTtl.let { tsBuilder.accessTokenTimeToLive(it) }
        ts.reuseRefreshTokens.let { tsBuilder.reuseRefreshTokens(it) }
        ts.refreshTokenTtl.let { tsBuilder.refreshTokenTimeToLive(it) }

        ts.accessTokenFormat.let { fmt ->
            when (fmt.lowercase()) {
                "self-contained", "jwt" -> tsBuilder.accessTokenFormat(OAuth2TokenFormat.SELF_CONTAINED)
                "reference" -> tsBuilder.accessTokenFormat(OAuth2TokenFormat.REFERENCE)
            }
        }
        builder.tokenSettings(tsBuilder.build())
    }
    return builder.build()
}
